core/templates/domain/exploration/read-only-exploration-backend-api.service.ts

```diff
@@ -17,153 +17,202 @@
-    // Maps previously loaded explorations to their IDs.
-    var _explorationCache = [];
-
-    var _fetchExploration = function(
-        explorationId, version, successCallback, errorCallback) {
-      var explorationDataUrl = _getExplorationUrl(explorationId, version);
-
-      $http.get(explorationDataUrl).then(function(response) {
-        var exploration = angular.copy(response.data);
-        if (successCallback) {
-          successCallback(exploration);
-        }
-      }, function(errorResponse) {
-        if (errorCallback) {
-          errorCallback(errorResponse.data);
-        }
-      });
-    };
-
-    var _isCached = function(explorationId) {
-      return _explorationCache.hasOwnProperty(explorationId);
-    };
-
-    var _getExplorationUrl = function(explorationId, version) {
-      if (version) {
-        return UrlInterpolationService.interpolateUrl(
-          EXPLORATION_VERSION_DATA_URL_TEMPLATE, {
-            exploration_id: explorationId,
-            version: String(version)
-          });
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+
+import { AppConstants } from 'app.constants';
+import { ClassifierBackendDict } from 'domain/classifier/classifier.model';
+import { ParamChangeBackendDict } from 'domain/exploration/ParamChangeObjectFactory';
+import { ParamSpecsBackendDict } from 'domain/exploration/ParamSpecsObjectFactory';
+import { StateObjectsBackendDict } from 'domain/exploration/StatesObjectFactory';
+import { UrlInterpolationService } from 'domain/utilities/url-interpolation.service';
+
+interface StateClassifierMappingBackendDict {
+  [state: string]: ClassifierBackendDict;
+}
+
+interface ReadOnlyExplorationBackendDict {
+  'init_state_name': string;
+  'param_changes': ParamChangeBackendDict[];
+  'param_specs': ParamSpecsBackendDict;
+  'states': StateObjectsBackendDict;
+  'title': string;
+  'language_code': string;
+  'objective': string;
+  'correctness_feedback_enabled': boolean;
+}
+
+interface FetchExplorationBackendResponse {
+  'can_edit': boolean;
+  'exploration': ReadOnlyExplorationBackendDict;
+  'exploration_id': string;
+  'is_logged_in': boolean;
+  'session_id': string;
+  'version': number;
+  'preferred_audio_language_code': string;
+  'state_classifier_mapping': StateClassifierMappingBackendDict;
+  'auto_tts_enabled': boolean;
+  'correctness_feedback_enabled': boolean;
+  'record_playthrough_probability': number;
+}
+
+interface ExplorationCache {
+  [explorationId: string]: FetchExplorationBackendResponse;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ReadOnlyExplorationBackendApiService {
+  private _explorationCache: ExplorationCache = {};
+
+  constructor(
+    private http: HttpClient,
+    private urlInterpolationService: UrlInterpolationService) {}
+
+  private _fetchExploration(
```

Comment:
Can we make this return a Promise instead? Then we can have callsites use .then instead of having this function accept success/error callbacks.

Reply:
Hmm, there are functions in this file that pass custom success and error callbacks here. So, I think it's fine

Reply:
They don't need to though, right?

Instead of `_fetchExploration('id', 1, customSuccess, customError)` it can just be `_fetchExploration('id', 1).then(customSuccess, customError)`

Reply:
Right, done.